### DISCUSSION TITLE
How do i run external program for translate on fedora 39

### DISCUSSION THREAD

**Discussion by mehmet-ercan** [xiaoyifang/goldendict-ng#1542] [General]
When i added below code to Program tab in Dictionaries, i got an error like that ; 

**program:** trans -e google -s fr -t en -show-original y -show-original-phonetics n -show-translation y -no-ansi -show-translation-phonetics n -show-prompt-message n -show-languages y -show-original-dictionary n -show-dictionary n -show-alternatives n %GDWORD%

**error code:** Query error: The program has returned exit code 255.

i am using fedora 39. when i used Ubuntu repo, i did not got any error, but i changed my repo to fedora, i am getting eror nowadays, thanks for help, regards. 

PS: i installed translate-shell program in link; https://github.com/soimort/translate-shell

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/7a00d412-92c9-433d-a699-f7780972c835)

**Comment by xiaoyifang** [comment]
What's the output of the program on console?

**Reply by mehmet-ercan** [reply]
actually I don't know where the output of the console, if you mean the output of the main window during translation I will show you, otherwise can you please show me?

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/e41a431d-9d14-41d7-8f71-0bd71e1d1718)

**Comment by xiaoyifang** [comment]
execute the command on the terminal

**Reply by mehmet-ercan** [reply]
okey, for example i wanna translate to apple en to fr, i ran below code on my terminal;

_command in golden-dict-ng->Programs tab_
trans -e google -s en -t fr -show-original y -show-original-phonetics n -show-translation y -no-ansi -show-translation-phonetics n -show-prompt-message n -show-languages y -show-original-dictionary n -show-dictionary n -show-alternatives n **%GDWORD%**

i ran in terminal for apple en to fr lang
**trans -e google -s en -t fr -show-original y -show-original-phonetics n -show-translation y -no-ansi -show-translation-phonetics n -show-prompt-message n -show-languages y -show-original-dictionary n -show-dictionary n -show-alternatives n apple**

the terminal output;
![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/3f43c4f2-e8ae-4f21-afdb-aec65e59b7d7)

**Reply by mehmet-ercan** [reply]
dear @xiaoyifang,
as you can see, the code runs on terminal perfectly, but does not run on goldendict-ng in Programs tab.
i dont know whats problem, maybe fedora's terminal, maybe goldendict-ng, or whatever?
thanks for all

**Comment by xiaoyifang** [comment]
https://github.com/soimort/translate-shell/issues/352

specify the  absolute path of `trans`

**Reply by mehmet-ercan** [reply]
thanks @xiaoyifang,  

i changed last line trans script with respect to https://github.com/soimort/translate-shell/issues/352,
then i added trans location in goldendict on Programs tab, 
but i got same error, i dont understand what is problem, maybe goldendict, maybe fedora's terminal or connection between goldendict and fedora's terminal or whatever, maybe tlink which you post me is according to macintosh os
the screenshots;

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/fab21cb5-0810-48b4-89c5-4ad8bed55b3f)

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/6dfebd9f-b7c9-45ba-9685-06aadf886c8c)
![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/abc51a91-f0c1-4bee-ab44-3f9010f164cd)
![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/e0bada3c-105d-4af5-8979-2282c88e4e9b)

**Comment by mehmet-ercan** [comment]
IT SHOULDNT BE HARD !!!

thanks to @shenlebantongying and @xiaoyifang for all help;

@shenlebantongying mailed me at 07.06.2024 about my goldendict working issue on fedora 39 like that;

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/3ede8a19-b9f9-4952-b205-b59344c5c14e)

Then I worked on the issue. When I searched with the same problem, I found some posts on GitHub, such as [this](https://github.com/soimort/translate-shell/issues/352) and [this](https://github.com/flathub/io.github.xiaoyifang.goldendict_ng/issues/1#issuecomment-1656071332).

But, even though I applied the code from the programs tab in GoldenDict according to the above posts, the result was the same—it didn't work.

![image](https://github.com/xiaoyifang/goldendict-ng/assets/112494004/5eec9dec-20d4-4de8-a037-d4867424589d)

Then I ran GoldenDict with that code in the terminal:
flatpak run --talk-name=org.freedesktop.Flatpak io.github.xiaoyifang.goldendict_ng

and tried translating a word—wow, it worked!

For more generic and faster usage, I created a bash script with flatpak run --... and created a desktop entry by linking that desktop file.

So, I achieved the goal, but it took about 3 weeks. Finally, I can say, "IT SHOULDN'T BE HARD!!!"

I think the problem might be related to Fedora's terminal and Flatpak. Maybe I'm not very knowledgeable about this subject or something else I don't know.

I hope for it has become like a document for those who come after me so that they will not experience the same problems.

**Reply by shenlebantongying** [reply]
I added some doc a few days ago 😅 
https://xiaoyifang.github.io/goldendict-ng/topic_flatpak/#add-external-programs

**Reply by mehmet-ercan** [reply]
Aha 😅, the document you prepared is very good, thank you @shenlebantongying
